### PR TITLE
Persistent identity mode (and annonymous mode)

### PR DESCRIPTION
**Pull request by marius-cristian**
Now we can all have a persistent identity and keep our devices 1 click connect :D
And if very annonymous we can also be fully anonymous with the same epehemral keys.
(like it was intendend in the beginning)

### CODE REVIEW

File: pkg/identity/identity.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+	if hostname == "" {
+		hostname = "unknown-host"
+	}
+	u, _ := user.Current()
+	username := "unknown-user"
+	if u != nil {
+		username = u.Username
+	}
+
+	ikm := []byte(hostname + ":" + username)
```
Comment (ac999) [inline pkg/identity/identity.go]: **Finding 1 — IKM for `identity.enc` falls outside the scope of NIST SP 800-56Cr2.**

`deriveFileKey` builds the HKDF input as `ikm := hostname + ":" + username` and runs HKDF-SHA512 with a fixed label (`pkg/identity/identity.go:184–198`, `pkg/crypto/asymmetric.go:269–273`).

NIST SP 800-56Cr2 §3.1 defines the input the recommendation covers:

> *"**Shared secret**: The secret byte string that is computed/generated during the execution of an **approved** key-establishment scheme and used as input to a key-derivation method as part of that transaction."*

§4 (one-step KDF preamble) makes the underlying PRF assumption explicit:

> *"Given knowledge of the key, k, and (perhaps) partial knowledge of a message x that includes an **unknown** substring y, it must be computationally infeasible to predict the (as-yet-unseen) value of MAC(k, x, …) with a probability of success that is a significant improvement over simply guessing either the MAC value or the value of y…"*

Hostname and username are public, often default, and inferable from the same data sources that would carry an exfiltrated `identity.enc`. They are not a "shared secret" by §3.1 and they are not the "unknown substring y" that §4's PRF assumption depends on.

A salt would not change that. SP 800-56Cr2 §3.1 defines salt as *"a byte string (which may be secret or non-secret) that is used as a MAC key"*, and §8.2 ("The Choice of a Salt Value") states:

> *"This Recommendation does not require the use of a randomly selected salt value. In particular, if there are no means to select a salt value and share it with all of the participants during a key-establishment transaction, then this Recommendation specifies that a predetermined default (e.g., all-zero) byte string be used as the salt value."*

Salt is the MAC key for the extraction step, not a substitute for entropy in `Z`.

The same construction is reused for `contacts.enc` (`pkg/identity/contacts.go:55, 78`).

Long-form companion with threat-model walkthrough and PoC posted separately on this PR.

**Review by ac999 — COMMENTED** [review]
## Verdict: Comment (not block)

Reviewed `git diff origin/main...origin/settings_screen_and_persistent_identity` end-to-end (+2018/−11, 18 files). Two findings worth surfacing before merge, three questions, a few nits. Inline comments below pin the specific lines. The full security write-up + PoC for Finding 1 is in a separate comment on this PR.

### Verified clean (no findings)

| Concern | Result | Evidence |
|---|---|---|
| `KD-2026-002` X25519 salt regression | ✅ Not regressed | `pkg/crypto/asymmetric.go` is +24/−0 — only adds `DeriveIdentityFileKey` and `DerivePresenceKey`. The X25519 ciphertext format from PR #31 is untouched. |
| Wire-protocol backward compatibility (new `Persistent` field) | ✅ Compatible | (a) Field is `json:"persistent,omitempty"`; (b) `grep -rn DisallowUnknownFields pkg/ cmd/` returns 0 — Go's permissive decoder ignores unknown fields. Old↔new peers handshake fine. |
| `OwnIsPersistent` / `PeerIsPersistent` aren't dead stores | ✅ Consumed | `git grep IsPersistent` shows reads at `pkg/logic/common/logic.go:1031` (`SaveCurrentPeerAsContact` gates on it) and `:1050` (accessor). |
| Anonymous mode still works | ✅ Yes | `cmd/kd/main.go:128–134` — `EnablePersistentIdentity` gated on `!cfg.Incognito`; ephemeral path unchanged. |
| File modes for sensitive files | ✅ `0600` for files, `0750` for dir | `pkg/identity/identity.go:84`, `contacts.go:73`. |
| HKDF domain separation | ✅ Distinct labels | `keibidrop-identity-file-v1` and `keibidrop-presence-v1`. |
| Test coverage on happy paths | ✅ Reasonable | Round-trip persistence, fingerprint stability, encrypted-not-plaintext, ML-KEM/X25519 round-trip, contacts CRUD. |
| `RelayEndoint` typo in `presence.go` | ✅ Pre-existing | `git grep RelayEndoint origin/main` shows hits in `pkg/logic/common/types.go:33`, `utils.go:97,162`. New code correctly uses the existing typo'd field. Worth fixing in a separate PR. |

### Findings (see inline comments for line-specific detail)

1. **Identity file key derivation has no real entropy.** Derives from `hostname + ":" + username` via HKDF — both are inferable, not secret. ChaCha20-Poly1305 itself is correct; the weakness is the KDF input. Real risk in the cloud-backup-leak scenario (file leaves the host along with usually-correlatable hostname/username clues). Full PoC + threat-model walkthrough in a separate comment on this PR. Recommended follow-up: add a per-install random salt at `~/.config/keibidrop/.salt`.
2. **Corrupted identity file requires manual filesystem intervention.** `LoadOrCreate` only retries on `os.IsNotExist`; any decrypt/AEAD failure (incl. legitimate "ran on a different machine after rename") leaves the broken file in place and silently falls back to ephemeral mode. Suggested: rename broken file to `identity.enc.broken-<ts>`, log a warning with the path, recreate.

### Questions

1. Is identity-binding to current hostname/username intentional (security anchor) or incidental (side effect of using machine data as KDF input)? If intentional, please document. If incidental, the salt fix from Finding 1 decouples it.
2. Should `serializedIdentity` (and `Contact`) carry a schema version field? `pkg/identity/identity.go:25–29` has no `Version int` — any future schema change needs a migration plan, and adding a version today is a one-line ergonomic win.
3. The new settings-screen UI adds literals not in the canonical palette: `#ccc` (×8), `#3b82f6`, `#60a5fa`, `#22c55e`, `#d97706`. (Note: `#888` ×9 is **in palette** — `#888` is shorthand for `#888888`, which is canonical.) Not a finding against this PR — codebase already drifts widely. I'm filing a separate issue proposing a centralized theme file; this just feeds into it.

### Nits (no action required)

1. `pkg/logic/common/presence.go:38` — `sendPresenceForAll` takes an inline interface for the logger. The rest of the codebase uses `kd.logger.With(...)`. Pure style.
2. `pkg/identity/identity.go:185–193` — `os.Hostname()` and `user.Current()` errors are silently swallowed and replaced with `"unknown-host"` / `"unknown-user"`. Edge case where a user later resolves their hostname shifts the file key.

### What I did NOT run locally

- `go test ./pkg/identity/... -count=1 -v` and `make build-static-rust-bridge && (cd rust && cargo build --release)` — deferred to avoid disturbing the working tree this session. CI presumably covers them. Worth a green tick before merge.

File: pkg/identity/identity.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+// LoadOrCreate loads an existing identity from configDir, or creates and
+// persists a new one if none exists.
+func LoadOrCreate(configDir string) (*DeviceIdentity, error) {
+	id, err := Load(configDir)
+	if err == nil {
+		return id, nil
+	}
+	if !os.IsNotExist(err) {
+		return nil, fmt.Errorf("load identity: %w", err)
```
Comment (ac999) [inline pkg/identity/identity.go:52]: **Finding 2 — robustness: corrupted identity file is a manual-recovery dead end.**

`LoadOrCreate` only retries when the file doesn't exist. Any *other* error path (decrypt failure from corruption, AEAD mismatch from running on a different machine after a hostname/user change, partial write from power loss, permission flip) bubbles up and `EnablePersistentIdentity` returns the error. `cmd/kd/main.go:130–133` then logs a warning and falls back to ephemeral mode, but the broken file is left in place — subsequent runs hit the same error every time. The user's only recovery path is to manually `rm ~/.config/keibidrop/identity.enc`, and there's no log message pointing at the path.

The test suite covers the happy paths well but nothing for: decrypt-fail, truncation, wrong-machine load.

**Suggested fix:** on a `Load` failure that is NOT `os.IsNotExist`, rename the broken file to `identity.enc.broken-<RFC3339-ts>`, log a warning with the path, then fall through to the create branch.

```go
if !os.IsNotExist(err) {
    backupPath := path + ".broken-" + time.Now().UTC().Format(time.RFC3339)
    _ = os.Rename(path, backupPath)
    logger.Warn("identity.enc was unreadable; preserved at",
        "path", backupPath, "error", err)
    // fall through to create
}
```

Plus a test that writes garbage to `identity.enc`, calls `LoadOrCreate`, and asserts: rename happened, new identity created, fingerprint differs from a previous run. Same applies to `contacts.enc` in `contacts.go`.

File: pkg/identity/identity.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+type DeviceIdentity struct {
+	Fingerprint string
+	Keys        *kbc.OwnKeys
+	CreatedAt   time.Time
+}
+
+// serializedIdentity is the JSON-serializable form of DeviceIdentity.
+// Private key seeds are stored; public keys are derived on load.
+type serializedIdentity struct {
```
Comment (ac999) [inline pkg/identity/identity.go:37]: **Q2 — forward-compat: should `serializedIdentity` carry a schema version?**

No `Version int` field on `serializedIdentity` (here) or `Contact` (in `contacts.go:21`). Any future schema addition (Ed25519, expiry, profile metadata, key rotation timestamp) will need migration logic without an anchor to key off. Adding a version field today is a one-line ergonomic win:

```go
type serializedIdentity struct {
    Version    int       `json:"v"`            // 1 today; bump when schema changes
    X25519Seed []byte    `json:"x25519_seed"`
    MLKEMSeed  []byte    `json:"mlkem_seed"`
    CreatedAt  time.Time `json:"created_at"`
}
```

Then on `Load`, switch on `s.Version` and dispatch to a migrator. For v1, the migrator is just identity. Adding it now is much cheaper than retrofitting once a schema-2 file exists in the wild.

Same suggestion for `Contact` — bonus benefit there is that you can evolve the `LastSeen` semantics or add per-contact key history without breaking older clients.

Not a merge-blocker. Easy to land before merge or as a small follow-up.

**Comment by ac999** [comment]
Walkthrough and PoC for the inline finding on `pkg/identity/identity.go:188`. Posted separately because it does not fit on a single line.

---

## TL;DR

`~/.config/keibidrop/identity.enc` is encrypted with ChaCha20-Poly1305 using a key derived via HKDF-SHA512 from `hostname + ":" + username` with a fixed label (`keibidrop-identity-file-v1`). No salt, no per-install secret, no user passphrase, no OS-keychain integration. The same construction is reused for `contacts.enc` (`pkg/identity/contacts.go:55, 78`).

NIST SP 800-56Cr2 §2 ("Scope and Purpose"):

> *"This Recommendation specifies two categories of key-derivation methods that can be employed, as required, to derive keying material from a shared secret Z generated during the execution of a key-establishment scheme specified in [SP 800-56A] or [SP 800-56B]."*

§3.1 ("Shared secret"):

> *"The secret byte string that is computed/generated during the execution of an **approved** key-establishment scheme and used as input to a key-derivation method as part of that transaction."*

`hostname + ":" + username` is neither — not the output of an approved key-establishment scheme, and not secret. The construction sits outside the scope SP 800-56Cr2 covers.

---

## What the construction is

`pkg/identity/identity.go:184–198`:

```go
func deriveFileKey() ([]byte, error) {
    hostname, _ := os.Hostname()
    if hostname == "" {
        hostname = "unknown-host"
    }
    u, _ := user.Current()
    username := "unknown-user"
    if u != nil {
        username = u.Username
    }

    ikm := []byte(hostname + ":" + username)
    return kbc.DeriveIdentityFileKey(ikm)
}
```

`pkg/crypto/asymmetric.go:269–273`:

```go
func DeriveIdentityFileKey(machineEntropy []byte) ([]byte, error) {
    h := sha512.Sum512(machineEntropy)
    return deriveKeyInternal(sha512.New, "keibidrop-identity-file-v1", KeySize, h[:])
}
```

The 32-byte key is a deterministic function of two strings. HKDF expansion adds no randomness — it is a key-derivation function, not a key-generation function.

---

## What NIST SP 800-56Cr2 says about this construction

§4 (one-step key-derivation preamble) states the PRF assumption that the security argument for these KDFs rests on:

> *"Given knowledge of the key, k, and (perhaps) partial knowledge of a message x that includes an **unknown** substring y, it must be computationally infeasible to predict the (as-yet-unseen) value of MAC(k, x, …) with a probability of success that is a significant improvement over simply guessing either the MAC value or the value of y, even if one has already seen the values of MAC(k_j, x_j, …) for a feasible number of other (k_j, x_j) pairs where each key k_j is known and each (partially known) message x_j includes the same unknown substring y, provided that none of the (k_j, x_j) pairs is identical to (k, x)."*

The "unknown substring `y`" is `Z`. The assumption is that `Z` is unknown to the adversary.

§3.1 also defines `salt`:

> *"a byte string (which may be secret or non-secret) that is used as a MAC key by either: 1) a MAC-based auxiliary function H employed in one-step key derivation or 2) a MAC employed in the randomness-extraction step during two-step key derivation."*

§8.2 ("The Choice of a Salt Value"):

> *"This Recommendation does not require the use of a randomly selected salt value. In particular, if there are no means to select a salt value and share it with all of the participants during a key-establishment transaction, then this Recommendation specifies that a predetermined default (e.g., all-zero) byte string be used as the salt value."*

Salt — random or not, present or absent — is the MAC key for the extraction step, not a secret substitute for `Z`.

---

## End-to-end exploit chain

### Phase 1 — file exfiltration

The attacker needs `~/.config/keibidrop/identity.enc` (and optionally `contacts.enc`). Realistic paths:

1. **Cloud backup leak.** User syncs `~/.config/` to Dropbox / iCloud Drive / Google Drive / OneDrive. The cloud account is compromised (credential reuse, phishing, stolen session token). Encrypted blobs are useful protection only when the KDF input is genuinely secret — the LastPass vault leak is the precedent worth keeping in mind.
2. **Stolen device, no FDE.** The encrypted file lives on a laptop disk. Without FDE, the file is readable by anyone with physical access.
3. **Malicious local process.** Any process running as the user can `read("identity.enc")` and call `os.Hostname()` / `whoami`. At this severity, KeibiDrop's identity is one of many compromised secrets, but it is still leaked.
4. **Forensic seizure.** Subpoena, customs, etc. State actor with file access.
5. **Misconfigured backup software.** Time Machine on a network share with weak ACLs; `rsync` to a shared NAS; etc.

For paths 1, 2, 4, 5 the file leaves the host along with a lot of *other* user data (shell history, gitconfig, ssh known_hosts, browser profiles). Hostname and username are typically right there in the same backup.

### Phase 2 — recover hostname / username

If the attacker has any of the following from the same backup or breach, they have the KDF input:

- `~/.bash_history` mentions `whoami` output, `ssh user@host`, etc.
- `~/.gitconfig` has the user's name (often matches Linux username for solo dev laptops)
- `~/.ssh/known_hosts` includes the host's own SSH host key entry on some setups
- The backup filename / path commonly encodes them: `andrei-laptop-backup-2026-04-30.tar.gz`, `iCloud/MacBook Pro/...`
- macOS-specific: `~/Library/Preferences` has the host name embedded in many plists
- `/etc/hostname` is in the backup if it's a full-disk image

Even without other context, brute force is trivial:

- Hostname space: ~10⁴–10⁶ realistic candidates per target. Default forms: `<firstname>-MacBook-Pro`, `<firstname>-laptop`, `<firstname>-pc`, `MacBook Pro` (the Apple default), distro defaults like `ubuntu`, `arch`, `debian`. Dictionaries of common patterns, plus the user's first name from public sources.
- Username space: ~10²–10³ — usually the user's GitHub login, email prefix, or `<firstname>` directly.
- Combined: ~10⁶–10⁹ candidates.

Each candidate requires one ChaCha20-Poly1305 verification on the file's first block. On a single modern x86 CPU core, ~10⁷ Poly1305 verifications/second is achievable without specialized hardware. Worst case ~100 CPU-seconds; typical case (informed guesses) <1 second.

### Phase 3 — exploit the recovered identity

Once `identity.enc` decrypts, the attacker has the JSON:

```json
{
  "x25519_seed": "<32-byte seed>",
  "mlkem_seed": "<seed>",
  "created_at": "..."
}
```

These seeds reconstruct the user's X25519 and ML-KEM private keys — the persistent identity that the feature is built around.

What the attacker can do with the keys:

1. **Impersonate the user.** Spin up a KeibiDrop instance with these keys. Any peer the user has shared a fingerprint with — i.e. anyone in `contacts.enc`, plus anyone the user previously ran a "1-click connect" with — sees the attacker's session as the legitimate user. The peer has no way to detect the swap; the fingerprint matches.
2. **Recover the contact graph.** `contacts.enc` decrypts with the same key. Attacker learns who the user talks to.
3. **Receive files intended for the user.** Race-or-replace: if a contact tries to "1-click connect" to the user, the attacker's listener may answer first. Files destined for the user land at the attacker.
4. **Send malicious files as the user.** A contact receives a malicious payload from what looks like a trusted peer. The trust model the feature is built on collapses into "first-mover wins".

There is no revocation path. The user has no signal that compromise has happened short of noticing the contact graph or files going somewhere unexpected.

---

## Proof of concept

The PoC is small because the construction is fully public — no novel cryptanalysis is needed. The attacker calls the same KDF the legitimate code uses.

```go
// poc.go — decrypt a stolen identity.enc given hostname/username guesses.
//
// Usage:
//   go run poc.go path/to/identity.enc <hostname> <username>
//
// Or for brute force, wrap the inner block in two for-loops over candidate lists.
//
// This file is intentionally simple — no novel crypto, no side-channel tricks.
// It only re-uses the same exposed helpers the legitimate code uses.

package main

import (
    "bufio"
    "fmt"
    "os"
    "strings"

    kbc "github.com/KeibiSoft/KeibiDrop/pkg/crypto"
)

func tryDecrypt(ciphertext []byte, hostname, username string) ([]byte, bool) {
    ikm := []byte(hostname + ":" + username)
    key, err := kbc.DeriveIdentityFileKey(ikm)
    if err != nil {
        return nil, false
    }
    pt, err := kbc.Decrypt(key, ciphertext)
    if err != nil {
        return nil, false
    }
    return pt, true
}

func main() {
    if len(os.Args) < 2 {
        fmt.Fprintln(os.Stderr, "usage: poc <identity.enc> [hostname username]")
        os.Exit(2)
    }
    ct, err := os.ReadFile(os.Args[1])
    if err != nil {
        panic(err)
    }

    // Mode 1 — single guess (when attacker already knows host/user).
    if len(os.Args) == 4 {
        if pt, ok := tryDecrypt(ct, os.Args[2], os.Args[3]); ok {
            fmt.Println(string(pt))
            return
        }
        fmt.Fprintln(os.Stderr, "decrypt failed: wrong host/user, or not an identity file")
        os.Exit(1)
    }

    // Mode 2 — brute force from stdin lines of "host:user".
    fmt.Fprintln(os.Stderr, "reading 'host:user' candidates from stdin…")
    sc := bufio.NewScanner(os.Stdin)
    for sc.Scan() {
        line := strings.TrimSpace(sc.Text())
        i := strings.IndexByte(line, ':')
        if i < 0 {
            continue
        }
        if pt, ok := tryDecrypt(ct, line[:i], line[i+1:]); ok {
            fmt.Println("HIT:", line)
            fmt.Println(string(pt))
            return
        }
    }
    fmt.Fprintln(os.Stderr, "no candidates matched")
    os.Exit(1)
}
```

Brute-force input via a Python one-liner:

```bash
python3 -c '
hosts = ["MacBook-Pro", "MacBookPro", "andrei-laptop", "andrei-pc", "ubuntu", "arch",
         "andreis-MacBook-Pro", "andrei-MacBook-Pro", "iMac", "Andreis-Mac"]
users = ["andrei", "andreicristian097", "ac", "user", "ubuntu", "ac999"]
for h in hosts:
    for u in users:
        print(f"{h}:{u}")
' | go run poc.go path/to/identity.enc
```

For a known target, the candidate list is usually <50 strings and runs in milliseconds. The PoC uses `kbc.DeriveIdentityFileKey` and `kbc.Decrypt` directly — no need to reimplement anything.

---

## Threat-model table

| Scenario | File obtained? | host+user known? | Decryption feasible? |
|---|---|---|---|
| Casual file inspection by a curious roommate | usually no (0600 perms) | yes | n/a |
| Stolen device with full-disk encryption on | no | yes | n/a |
| Stolen device, no FDE | yes | yes | yes |
| Cloud-backup credential leak (single vendor) | yes | likely yes | yes |
| Targeted phishing where attacker convinces user to send `~/.config/` | yes | yes | yes |
| Compromised local process running as user | yes | yes | yes |
| Network adversary (no file access) | no | maybe | n/a |
| Forensic / state actor with seized device | yes | yes | yes |

The cloud-backup-leak row is the cleanest path — a single vendor compromise, no other prerequisites. Same shape as the LastPass vault leak: encrypted blobs are useful protection only when the KDF input is genuinely secret.

---

## Industry-baseline comparison

| Tool | At-rest protection of long-lived keys |
|---|---|
| OpenSSH (default) | plaintext private key, file mode 0600 |
| GPG | passphrase-encrypted via OpenPGP S2K |
| 1Password / Bitwarden / KeePassXC | user master password + Argon2id / PBKDF2-SHA256 |
| macOS Keychain Services | hardware-backed (Secure Enclave on supported hardware) |
| Linux libsecret (gnome-keyring / kwallet) | unlocked at login by user password |
| Windows DPAPI | bound to user's Windows password / TPM |
| Signal Desktop (pre-2018 fix) | similar shape; remediated after public disclosure |

The Signal Desktop precedent (2018) is worth flagging — a comparable construction shape, treated as a real vulnerability and remediated.

---

## False-positive check

The PoC is ~30 lines that call this codebase's own public functions. If the build succeeds for this PR, the build succeeds for the PoC.

---

## Where the standards point

NIST SP 800-56Cr2's scope (Abstract, §2) is keying material derived from a shared secret produced by SP 800-56A or SP 800-56B key establishment.

For the at-rest protection of long-lived secrets when no in-band shared secret is available, NIST treats those as different problems and points at different documents:

- **SP 800-132** — *Recommendation for Password-Based Key Derivation* — covers derivation of cryptographic keys from a user-supplied password using PBKDF2 with a random salt and a tunable iteration count.
- **SP 800-108** — *Recommendation for Key Derivation Using Pseudorandom Functions* — referenced by SP 800-56Cr2 §5.1 for the key-expansion step. Also assumes the input key has full entropy.

This is recorded for the scope-of-applicability question, not as a recommendation for what KeibiDrop should do.

---

## One-line summary

`identity.enc` and `contacts.enc` derive their AEAD key via HKDF-SHA512 from `hostname + ":" + username`. Per NIST SP 800-56Cr2 §3.1, the IKM (`Z`) the recommendation covers is required to be a shared secret produced by an approved key-establishment scheme; per §8.2, the salt (currently absent) is not a substitute for that secrecy.

**Comment by marius-cristian** [comment]
Offotopic: Big headache from too much info. Keep it short and on point. Dont need to read a 10 page novels with artefacts. Same with comments. Same with introducing new error types.
